runtime/parachains/src/paras_inherent.rs

```diff
@@ -321,19 +332,33 @@ pub mod pallet {
 			data: ParachainsInherentData<T::Header>,
 		) -> DispatchResultWithPostInfo {
 			let ParachainsInherentData {
-				bitfields: signed_bitfields,
-				backed_candidates,
+				bitfields: mut signed_bitfields,
+				mut backed_candidates,
 				parent_header,
-				disputes,
+				mut disputes,
 			} = data;
-			let total_weight =
-				paras_inherent_total_weight::<T>(&backed_candidates, &signed_bitfields, &disputes);
+
+			let mut candidate_weight = backed_candidates_weight::<T>(&backed_candidates);
+			let mut bitfields_weight = signed_bitfields_weight::<T>(signed_bitfields.len());
+			let disputes_weight = dispute_statements_weight::<T>(&disputes);
+
+			let max_block_weight = <T as frame_system::Config>::BlockWeights::get().max_block;
 
 			// Abort if the total weight of the block exceeds the max block weight
-			ensure!(
-				total_weight <= <T as frame_system::Config>::BlockWeights::get().max_block,
-				Error::<T>::InherentOverweight
-			);
+			if candidate_weight.saturating_add(bitfields_weight).saturating_add(disputes_weight) > max_block_weight {
+				backed_candidates.clear();
+				candidate_weight = 0;
+				signed_bitfields.clear();
+				bitfields_weight = 0;
+			}
+
+			let total_weight = if disputes_weight > max_block_weight {
+				let parent_hash = <frame_system::Pallet<T>>::parent_hash();
+				let entropy = compute_entropy::<T>(parent_hash);
```

Comment:
we can compute the entropy once at the top of enter so we don't hav to compute it multiple times

Reply:
We should create the `Rng` only once and re-use that.

Reply:
i am gonna make a little pr here with this refactor rn

Reply:
https://github.com/paritytech/polkadot/pull/4287